src/util/SignalsHandler.cpp

```diff
@@ -0,0 +1,130 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/SignalsHandler.hpp"
+
+#include "util/Assert.hpp"
+#include "util/Constants.hpp"
+#include "util/config/Config.hpp"
+#include "util/log/Logger.hpp"
+
+#include <boost/asio/error.hpp>
+#include <boost/asio/executor_work_guard.hpp>
+
+#include <chrono>
+#include <cmath>
+#include <csignal>
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <string>
+#include <utility>
+
+namespace util {
+
+namespace impl {
+
+class SignalsHandlerStatic {
+    static SignalsHandler* handler_;
+
+public:
+    static void
+    registerHandler(SignalsHandler& handler)
+    {
+        ASSERT(handler_ == nullptr, "There could be only one instance of SignalsHandler");
+        handler_ = &handler;
+    }
+
+    static void
+    resetHandler()
+    {
+        handler_ = nullptr;
+    }
+
+    static void
+    handleSignal(int signal)
+    {
+        ASSERT(handler_ != nullptr, "SignalsHandler is not initialized");
+        handler_->stopHandler_(signal);
+    }
+
+    static void
+    handleSecondSignal(int signal)
+    {
+        ASSERT(handler_ != nullptr, "SignalsHandler is not initialized");
+        handler_->secondSignalHandler_(signal);
+    }
+};
+
+SignalsHandler* SignalsHandlerStatic::handler_ = nullptr;
+
+}  // namespace impl
+
+SignalsHandler::SignalsHandler(Config const& config, std::function<void(std::string)> forceExitHandler)
```

Comment:
I am thinking the type of forceExitHandler std::function<void(std::string)> . Do we really need std::string here? It seems only for logging. How about log in lambda ?

Reply:
Good idea, thanks!